proparse/src/main/java/org/prorefactor/proparse/antlr4/MultiChannelTokenSource.java

```diff
@@ -0,0 +1,96 @@
+/*******************************************************************************
+ * Copyright (c) 2003-2015 John Green
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    John Green - initial API and implementation and/or initial documentation
+ *******************************************************************************/ 
+package org.prorefactor.proparse.antlr4;
+
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.IntStream;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.TokenFactory;
+import org.antlr.v4.runtime.TokenSource;
+import org.antlr.v4.runtime.WritableToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Last layer of TokenSource, used to dispatch preprocessor statements, comments and whitespaces to the right channel 
+ */
+public class MultiChannelTokenSource implements TokenSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MultiChannelTokenSource.class);
+
+  // All preprocessor statements (&MESSAGE, &ANALYZE-SUSPEND and RESUME, &GLOBAL/SCOPED DEFINE and &UNDEFINE) go to this channel
+  public static final int PREPROCESSOR_CHANNEL = 2;
+  // All &_PROPARSE statements go to this channel
+  public static final int PROPARSE_CHANNEL = 3;
+  
+  private final TokenSource source;
+  private Token currentToken;
+
+  public MultiChannelTokenSource(TokenSource input) {
+    this.source = input;
+  }
+
+  @Override
+  public Token nextToken() {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "nextToken" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)